File: Wrp/python/src/com/amivoice/wrp/Wrp.py
import sys
import threading
import time

try:
	from urllib import quote
	def u(s):
		if isinstance(s, str):
			return s.decode(sys.stdin.encoding)
		return s
	def q(s):
		return quote(s.encode("UTF-8"))
except:
	from urllib.parse import quote
	def u(s):
		return s
	def q(s):
		return quote(s, "UTF-8")

class Wrp(object):
	VERSION = "Wrp/1.0.03 Python/%d.%d.%d (%s)" % (sys.version_info[0], sys.version_info[1], sys.version_info[2], sys.platform)

	@staticmethod
	def getVersion():
		return Wrp.VERSION

	@staticmethod
	def construct(implementation = 1):
		if implementation == 1:
			from .Wrp_ import Wrp_
			return Wrp_()
		else:
			raise ValueError("Unknown implementation: " + implementation)

	def __init__(self):
		self.condition_ = threading.Condition()
		self.listener_ = None
		self.serverURL_ = None
		self.proxyServerName_ = None
		self.connectTimeout_ = 0
		self.receiveTimeout_ = 0
		self.grammarFileNames_ = None
		self.profileId_ = None
		self.profileWords_ = None
		self.segmenterProperties_ = None
		self.keepFillerToken_ = None
		self.resultUpdatedInterval_ = None
		self.extension_ = None
		self.authorization_ = None
		self.codec_ = None
		self.resultType_ = None
		self.state_ = 0
		self.waitingResults_ = 0

	def setListener(self, listener):
		self.listener_ = listener

	def setServerURL(self, serverURL):
		self.serverURL_ = u(serverURL)
		if self.serverURL_ != None:
			if self.serverURL_.startswith("http://"):
				self.serverURL_ = "ws://" + self.serverURL_[7:]
			elif self.serverURL_.startswith("https://"):
				self.serverURL_ = "wss://" + self.serverURL_[8:]

	def setProxyServerName(self, proxyServerName):
		self.proxyServerName_ = proxyServerName

	def setConnectTimeout(self, connectTimeout):
		self.connectTimeout_ = connectTimeout

	def setReceiveTimeout(self, receiveTimeout):
		self.receiveTimeout_ = receiveTimeout

	def setGrammarFileNames(self, grammarFileNames):
		self.grammarFileNames_ = u(grammarFileNames)

	def setProfileId(self, profileId):
		self.profileId_ = u(profileId)

	def setProfileWords(self, profileWords):
		self.profileWords_ = u(profileWords)

	def setSegmenterProperties(self, segmenterProperties):
		self.segmenterProperties_ = u(segmenterProperties)

	def setKeepFillerToken(self, keepFillerToken):
		self.keepFillerToken_ = keepFillerToken

	def setExtension(self, extension):
		self.extension_ = u(extension)

	def setAuthorization(self, authorization):
		self.authorization_ = u(authorization)

	def setCodec(self, codec):
		self.codec_ = u(codec)

	def setResultType(self, resultType):
		self.resultType_ = u(resultType)

	def setServiceAuthorization(self, serviceAuthorization):
		if serviceAuthorization != None:
			self.authorization_ = u(serviceAuthorization)

	def connect(self):
		if self.isConnected_():
			self.lastMessage_ = "WARNING: already connected to WebSocket server"
			if self.listener_ != None:
				self.listener_.TRACE(self.lastMessage_)
			return False
		if self.serverURL_ == None:
			self.lastMessage_ = "ERROR: can't connect to WebSocket server (Missing server URL)"
			if self.listener_ != None:
				self.listener_.TRACE(self.lastMessage_)
			return False
		try:
			self.state_ = 0
			self.waitingResults_ = 0
			self.connect_(self.serverURL_, self.proxyServerName_, self.connectTimeout_, self.receiveTimeout_)
		except Exception as e:
			self.lastMessage_ = "ERROR: can't connect to WebSocket server (" + u(str(e)) + "): " + self.serverURL_
			if self.listener_ != None:
				self.listener_.TRACE(self.lastMessage_)
			self.disconnect_()
			return False
		self.lastMessage_ = "INFO: connected to WebSocket server: " + self.serverURL_
		if self.listener_ != None:
			self.listener_.TRACE(self.lastMessage_)
		return True

	def connect_(self, serverURL, proxyServerName, connectTimeout, receiveTimeout):
		raise NotImplementedError()

	def disconnect(self):
		if not self.isConnected_():
			self.lastMessage_ = "WARNING: already disconnected from WebSocket server"
			if self.listener_ != None:
				self.listener_.TRACE(self.lastMessage_)
			return False
		self.disconnect_()
		self.lastMessage_ = "INFO: disconnected from WebSocket server"
		if self.listener_ != None:
			self.listener_.TRACE(self.lastMessage_)
		return True

	def disconnect_(self):
		raise NotImplementedError()

	def isConnected(self):
		return self.isConnected_()

	def isConnected_(self):
		raise NotImplementedError()

	def feedDataResume(self):
		with self.condition_:
			if not self.isConnected_():
				self.lastMessage_ = "WARNING: already disconnected from WebSocket server"
				if self.listener_ != None:
					self.listener_.TRACE(self.lastMessage_)
				return False
			if self.state_ != 0:
				self.lastMessage_ = "WARNING: already started feeding data to WebSocket server"
				if self.listener_ != None:
					self.listener_.TRACE(self.lastMessage_)
				return False
			command = ""
			command += "s "
			if self.codec_ != None and len(self.codec_) > 0:
				command += self.codec_
			else:
				command += "16K"
			if self.grammarFileNames_ != None and len(self.grammarFileNames_) > 0:
				command += ' '
				command += self.grammarFileNames_
				if self.grammarFileNames_.find('\001') != -1 and not self.grammarFileNames_.endswith("\001"):
					command += '\001'
			else:
				command += " \001"
			if self.profileId_ != None:
				command += " profileId="
				if self.profileId_.find(' ') != -1:
					command += '"'
					command += self.profileId_
					command += '"'
				else:
					command += self.profileId_
			if self.profileWords_ != None:
				command += " profileWords="
				if self.profileWords_.find(' ') != -1:
					command += '"'
					command += self.profileWords_
					command += '"'
				else:
					command += self.profileWords_
			if self.segmenterProperties_ != None:
				command += " segmenterProperties="
				if self.segmenterProperties_.find(' ') != -1:
					command += '"'
					command += self.segmenterProperties_
					command += '"'
				else:
					command += self.segmenterProperties_
			if self.keepFillerToken_ != None:
				command += " keepFillerToken="
				if self.keepFillerToken_.find(' ') != -1:
					command += '"'
					command += self.keepFillerToken_
					command += '"'
				else:
					command += self.keepFillerToken_
			if self.resultUpdatedInterval_ != None:
				command += " resultUpdatedInterval="
				if self.resultUpdatedInterval_.find(' ') != -1:
					command += '"'
					command += self.resultUpdatedInterval_
					command += '"'
				else:
					command += self.resultUpdatedInterval_
			if self.extension_ != None:
				command += " extension="
				if self.extension_.find(' ') != -1:
					command += '"'
					command += self.extension_.replace("\"", "\"\"")
					command += '"'
				else:
					command += self.extension_
			if self.authorization_ != None:
				command += " authorization="
				if self.authorization_.find(' ') != -1:
					command += '"'
					command += self.authorization_
					command += '"'
				else:
					command += self.authorization_
			if self.resultType_ != None:
				command += " resultType="
				if self.resultType_.find(' ') != -1:
					command += '"'
					command += self.resultType_
					command += '"'
				else:
					command += self.resultType_
			try:
				self.state_ = 1
				data = command.encode("UTF-8")
				self.sendRequest_(data, 0, len(data), 0)
				while self.state_ == 1:
					self.condition_.wait()
				if self.state_ == 5:
					return False
			except InterruptedError:
				self.lastMessage_ = "WARNING: interrupted to start feeding data to WebSocket server"
				if self.listener_ != None:
					self.listener_.TRACE(self.lastMessage_)
				return False
			except (IOError, OSError) as e:
				self.lastMessage_ = "ERROR: can't start feeding data to WebSocket server (" + u(str(e)) + ")"
				if self.listener_ != None:
					self.listener_.TRACE(self.lastMessage_)
				return False
			return True

	def feedData(self, data, dataOffset, dataBytes):
		with self.condition_:
			if not self.isConnected_():
				self.lastMessage_ = "WARNING: already disconnected from WebSocket server"
				if self.listener_ != None:
					self.listener_.TRACE(self.lastMessage_)
				return False
			if self.state_ != 2 and self.state_ != 3:
				self.lastMessage_ = "WARNING: already stopped feeding data to WebSocket server"
				if self.listener_ != None:
					self.listener_.TRACE(self.lastMessage_)
				return False
			try:
				self.state_ = 3
				self.sendRequest_(data, dataOffset, dataBytes, 0x70) # 'p'
			except (IOError, OSError) as e:
				self.lastMessage_ = "ERROR: can't feed data to WebSocket server (" + u(str(e)) + ")"
				if self.listener_ != None:
					self.listener_.TRACE(self.lastMessage_)
				return False
			return True

	def feedDataPause(self):
		with self.condition_:
			if not self.isConnected_():
				self.lastMessage_ = "WARNING: already disconnected from WebSocket server"
				if self.listener_ != None:
					self.listener_.TRACE(self.lastMessage_)
				return False
			if self.state_ != 2 and self.state_ != 3:
				self.lastMessage_ = "WARNING: already stopped feeding data to WebSocket server"
				if self.listener_ != None:
					self.listener_.TRACE(self.lastMessage_)
				return False
			try:
				self.state_ = 4
				data = b"e"
				self.sendRequest_(data, 0, len(data), 0)
				while self.state_ == 4:
					self.condition_.wait()
				if self.state_ == 5:
					return False
			except InterruptedError:
				self.lastMessage_ = "WARNING: interrupted to stop feeding data to WebSocket server"
				if self.listener_ != None:
					self.listener_.TRACE(self.lastMessage_)
				return False
			except (IOError, OSError) as e:
				self.lastMessage_ = "ERROR: can't stop feeding data to WebSocket server (" + u(str(e)) + ")"
				if self.listener_ != None:
					self.listener_.TRACE(self.lastMessage_)
				return False
			return True

	def getLastMessage(self):
		return self.lastMessage_

	def getWaitingResults(self):
		return self.waitingResults_

	def sleep(self, timeout):
		try:
			time.sleep(timeout / 1000.0)
		except:
			pass

	def sendRequest_(self, data, dataOffset, dataBytes, command):
		raise NotImplementedError()

	def onOpen_(self):
		pass

	def onClose_(self):
		pass

	def onError_(self, cause):
		with self.condition_:
			if self.state_ == 0 or self.state_ == 5:
				return
			self.lastMessage_ = "ERROR: caught exception (" + u(str(cause)) + ")"
			if self.listener_ != None:
				self.listener_.TRACE(self.lastMessage_)
			self.state_ = 5
			self.condition_.notifyAll()

	def onMessage_(self, message):
		if len(message) == 0:
			return
		command = message[0]
		if command == 's':
			if len(message) > 1:
				with self.condition_:
					self.lastMessage_ = "ERROR: can't start feeding data to WebSocket server (" + message[2:] + ")"
					if self.listener_ != None:
						self.listener_.TRACE(self.lastMessage_)
					self.state_ = 5
					self.condition_.notifyAll()
			else:
				with self.condition_:
					self.lastMessage_ = "INFO: started feeding data to WebSocket server"
					if self.listener_ != None:
						self.listener_.TRACE(self.lastMessage_)
					self.state_ = 2
					self.condition_.notifyAll()
		elif command == 'p':
			if len(message) > 1:
				with self.condition_:
					self.lastMessage_ = "ERROR: can't feed data to WebSocket server (" + message[2:] + ")"
					if self.listener_ != None:
						self.listener_.TRACE(self.lastMessage_)
					self.state_ = 5
					self.condition_.notifyAll()
			else:
				pass
		elif command == 'e':
			if len(message) > 1:
				with self.condition_:
					self.lastMessage_ = "ERROR: can't stop feeding data to WebSocket server (" + message[2:] + ")"
					if self.listener_ != None:
						self.listener_.TRACE(self.lastMessage_)
					self.state_ = 5
					self.condition_.notifyAll()
			else:
				with self.condition_:
					self.lastMessage_ = "INFO: stopped feeding data to WebSocket server"
					if self.listener_ != None:
						self.listener_.TRACE(self.lastMessage_)
					self.state_ = 0
					self.condition_.notifyAll()
		elif command == 'S':
			self.waitingResults_ += 1
			if self.listener_ != None:
				self.listener_.utteranceStarted(int(message[2:]))
		elif command == 'E':
			if self.listener_ != None:
				self.listener_.utteranceEnded(int(message[2:]))
		elif command == 'C':
			if self.listener_ != None:
				self.listener_.resultCreated()
		elif command == 'U':
			if self.listener_ != None:
				self.listener_.resultUpdated(message[2:])
		elif command == 'A':
			if self.listener_ != None:
				self.listener_.resultFinalized(message[2:])
			self.waitingResults_ -= 1
		elif command == 'R':
			if self.listener_ != None:
				self.listener_.resultFinalized("\001\001\001\001\001" + message[2:])
			self.waitingResults_ -= 1
		elif command == 'Q':
			if self.listener_ != None:
				self.listener_.eventNotified(command, message[2:])
		elif command == 'G':
			if self.listener_ != None:
				self.listener_.eventNotified(command, message[2:])

File: Wrp/cs/src/com/amivoice/wrp/Wrp.cs
using System;
using System.IO;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading;

#if !UNITY_2018_4_OR_NEWER
[assembly: AssemblyTitle("Wrp")]
[assembly: AssemblyProduct("Wrp")]
[assembly: AssemblyCopyright("Copyright (C) 2019-2021 Advanced Media, Inc.")]
[assembly: ComVisible(false)]
[assembly: Guid("d605a501-80ec-4f48-a8fb-2a70e627a256")]
[assembly: AssemblyVersion("1.0.03")]
[assembly: AssemblyFileVersion("1.0.03")]
#endif

namespace com.amivoice.wrp {

public abstract class Wrp {
	private static string VERSION = "Wrp/1.0.03 CLR/" + System.Environment.Version.ToString() + " (" + System.Environment.OSVersion.ToString() + ")";

	public static string getVersion() {
		return VERSION;
	}

	public static Wrp construct() {
		return new Wrp_();
	}

	public static Wrp construct(int implementation) {
		if (implementation == 1) {
			return new Wrp_();
		} else {
			throw new ArgumentException("Unknown implementation: " + implementation);
		}
	}

	private WrpListener listener_;
	private string serverURL_;
	private string proxyServerName_;
	private int connectTimeout_;
	private int receiveTimeout_;
	private string grammarFileNames_;
	private string profileId_;
	private string profileWords_;
	private string segmenterProperties_;
	private string keepFillerToken_;
	private string resultUpdatedInterval_;
	private string extension_;
	private string authorization_;
	private string codec_;
	private string resultType_;
	private int state_;
	private string lastMessage_;
	private volatile int waitingResults_;

	protected Wrp() {
		listener_ = null;
		serverURL_ = null;
		proxyServerName_ = null;
		connectTimeout_ = 0;
		receiveTimeout_ = 0;
		grammarFileNames_ = null;
		profileId_ = null;
		profileWords_ = null;
		segmenterProperties_ = null;
		keepFillerToken_ = null;
		resultUpdatedInterval_ = null;
		extension_ = null;
		authorization_ = null;
		codec_ = null;
		resultType_ = null;
		state_ = 0;
		waitingResults_ = 0;
	}

	public void setListener(WrpListener listener) {
		listener_ = listener;
	}

	public void setServerURL(string serverURL) {
		serverURL_ = serverURL;
		if (serverURL_ != null) {
			if (serverURL_.StartsWith("http://")) {
				serverURL_ = "ws://" + serverURL_.Substring(7);
			} else
			if (serverURL_.StartsWith("https://")) {
				serverURL_ = "wss://" + serverURL_.Substring(8);
			}
		}
	}

	public void setProxyServerName(string proxyServerName) {
		proxyServerName_ = proxyServerName;
	}

	public void setConnectTimeout(int connectTimeout) {
		connectTimeout_ = connectTimeout;
	}

	public void setReceiveTimeout(int receiveTimeout) {
		receiveTimeout_ = receiveTimeout;
	}

	public void setGrammarFileNames(string grammarFileNames) {
		grammarFileNames_ = grammarFileNames;
	}

	public void setProfileId(string profileId) {
		profileId_ = profileId;
	}

	public void setProfileWords(string profileWords) {
		profileWords_ = profileWords;
	}

	public void setSegmenterProperties(string segmenterProperties) {
		segmenterProperties_ = segmenterProperties;
	}

	public void setKeepFillerToken(string keepFillerToken) {
		keepFillerToken_ = keepFillerToken;
	}

	public void setResultUpdatedInterval(string resultUpdatedInterval) {
		resultUpdatedInterval_ = resultUpdatedInterval;
	}

	public void setExtension(string extension) {
		extension_ = extension;
	}

	public void setAuthorization(string authorization) {
		authorization_ = authorization;
	}

	public void setCodec(string codec) {
		codec_ = codec;
	}

	public void setResultType(string resultType) {
		resultType_ = resultType;
	}

	public void setServiceAuthorization(string serviceAuthorization) {
		if (serviceAuthorization != null) {
			authorization_ = serviceAuthorization;
		}
	}

	public bool connect() {
		if (isConnected_()) {
			lastMessage_ = "WARNING: already connected to WebSocket server";
			if (listener_ != null) {
				listener_.TRACE(lastMessage_);
			}
			return false;
		}
		if (serverURL_ == null) {
			lastMessage_ = "ERROR: can't connect to WebSocket server (Missing server URL)";
			if (listener_ != null) {
				listener_.TRACE(lastMessage_);
			}
			return false;
		}
		try {
			state_ = 0;
			waitingResults_ = 0;
			connect_(serverURL_, proxyServerName_, connectTimeout_, receiveTimeout_);
		} catch (Exception e) {
			lastMessage_ = "ERROR: can't connect to WebSocket server (" + e.Message + "): " + serverURL_;
			if (listener_ != null) {
				listener_.TRACE(lastMessage_);
			}
			disconnect_();
			return false;
		}
		lastMessage_ = "INFO: connected to WebSocket server: " + serverURL_;
		if (listener_ != null) {
			listener_.TRACE(lastMessage_);
		}
		return true;
	}

	protected abstract void connect_(string serverURL, string proxyServerName, int connectTimeout, int receiveTimeout);

	public bool disconnect() {
		if (!isConnected_()) {
			lastMessage_ = "WARNING: already disconnected from WebSocket server";
			if (listener_ != null) {
				listener_.TRACE(lastMessage_);
			}
			return false;
		}
		disconnect_();
		lastMessage_ = "INFO: disconnected from WebSocket server";
		if (listener_ != null) {
			listener_.TRACE(lastMessage_);
		}
		return true;
	}

	protected abstract void disconnect_();

	public bool isConnected() {
		return isConnected_();
	}

	protected abstract bool isConnected_();

	public bool feedDataResume() {
		lock (this) {
			if (!isConnected_()) {
				lastMessage_ = "WARNING: already disconnected from WebSocket server";
				if (listener_ != null) {
					listener_.TRACE(lastMessage_);
				}
				return false;
			}
			if (state_ != 0) {
				lastMessage_ = "WARNING: already started feeding data to WebSocket server";
				if (listener_ != null) {
					listener_.TRACE(lastMessage_);
				}
				return false;
			}
			StringBuilder command = new StringBuilder();
			command.Append("s ");
			if (codec_ != null && codec_.Length > 0) {
				command.Append(codec_);
			} else {
				command.Append("16K");
			}
			if (grammarFileNames_ != null && grammarFileNames_.Length > 0) {
				command.Append(' ');
				command.Append(grammarFileNames_);
				if (grammarFileNames_.IndexOf('\x01') != -1 && !grammarFileNames_.EndsWith("\x01")) {
					command.Append('\x01');
				}
			} else {
				command.Append(" \x01");
			}
			if (profileId_ != null) {
				command.Append(" profileId=");
				if (profileId_.IndexOf(' ') != -1) {
					command.Append('"');
					command.Append(profileId_);
					command.Append('"');
				} else {
					command.Append(profileId_);
				}
			}
			if (profileWords_ != null) {
				command.Append(" profileWords=");
				if (profileWords_.IndexOf(' ') != -1) {
					command.Append('"');
					command.Append(profileWords_);
					command.Append('"');
				} else {
					command.Append(profileWords_);
				}
			}
			if (segmenterProperties_ != null) {
				command.Append(" segmenterProperties=");
				if (segmenterProperties_.IndexOf(' ') != -1) {
					command.Append('"');
					command.Append(segmenterProperties_);
					command.Append('"');
				} else {
					command.Append(segmenterProperties_);
				}
			}
			if (keepFillerToken_ != null) {
				command.Append(" keepFillerToken=");
				if (keepFillerToken_.IndexOf(' ') != -1) {
					command.Append('"');
					command.Append(keepFillerToken_);
					command.Append('"');
				} else {
					command.Append(keepFillerToken_);
				}
			}
			if (resultUpdatedInterval_ != null) {
				command.Append(" resultUpdatedInterval=");
				if (resultUpdatedInterval_.IndexOf(' ') != -1) {
					command.Append('"');
					command.Append(resultUpdatedInterval_);
					command.Append('"');
				} else {
					command.Append(resultUpdatedInterval_);
				}
			}
			if (extension_ != null) {
				command.Append(" extension=");
				if (extension_.IndexOf(' ') != -1) {
					command.Append('"');
					command.Append(extension_.Replace("\"", "\"\""));
					command.Append('"');
				} else {
					command.Append(extension_);
				}
			}
			if (authorization_ != null) {
				command.Append(" authorization=");
				if (authorization_.IndexOf(' ') != -1) {
					command.Append('"');
					command.Append(authorization_);
					command.Append('"');
				} else {
					command.Append(authorization_);
				}
			}
			if (resultType_ != null) {
				command.Append(" resultType=");
				if (resultType_.IndexOf(' ') != -1) {
					command.Append('"');
					command.Append(resultType_);
					command.Append('"');
				} else {
					command.Append(resultType_);
				}
			}
			try {
				state_ = 1;
				byte[] data = Encoding.UTF8.GetBytes(command.ToString());
				sendRequest_(data, 0, data.Length, (byte)0);
				while (state_ == 1) {
					Monitor.Wait(this);
				}
				if (state_ == 5) {
					return false;
				}
			} catch (ThreadInterruptedException) {
				lastMessage_ = "WARNING: interrupted to start feeding data to WebSocket server";
				if (listener_ != null) {
					listener_.TRACE(lastMessage_);
				}
				return false;
			} catch (IOException e) {
				lastMessage_ = "ERROR: can't start feeding data to WebSocket server (" + e.Message + ")";
				if (listener_ != null) {
					listener_.TRACE(lastMessage_);
				}
				return false;
			}
			return true;
		}
	}

	public bool feedData(byte[] data, int dataOffset, int dataBytes) {
		lock (this) {
			if (!isConnected_()) {
				lastMessage_ = "WARNING: already disconnected from WebSocket server";
				if (listener_ != null) {
					listener_.TRACE(lastMessage_);
				}
				return false;
			}
			if (state_ != 2 && state_ != 3) {
				lastMessage_ = "WARNING: already stopped feeding data to WebSocket server";
				if (listener_ != null) {
					listener_.TRACE(lastMessage_);
				}
				return false;
			}
			try {
				state_ = 3;
				sendRequest_(data, dataOffset, dataBytes, (byte)'p');
			} catch (IOException e) {
				lastMessage_ = "ERROR: can't feed data to WebSocket server (" + e.Message + ")";
				if (listener_ != null) {
					listener_.TRACE(lastMessage_);
				}
				return false;
			}
			return true;
		}
	}

	public bool feedDataPause() {
		lock (this) {
			if (!isConnected_()) {
				lastMessage_ = "WARNING: already disconnected from WebSocket server";
				if (listener_ != null) {
					listener_.TRACE(lastMessage_);
				}
				return false;
			}
			if (state_ != 2 && state_ != 3) {
				lastMessage_ = "WARNING: already stopped feeding data to WebSocket server";
				if (listener_ != null) {
					listener_.TRACE(lastMessage_);
				}
				return false;
			}
			try {
				state_ = 4;
				byte[] data = Encoding.ASCII.GetBytes("e");
				sendRequest_(data, 0, data.Length, (byte)0);
				while (state_ == 4) {
					Monitor.Wait(this);
				}
				if (state_ == 5) {
					return false;
				}
			} catch (ThreadInterruptedException) {
				lastMessage_ = "WARNING: interrupted to stop feeding data to WebSocket server";
				if (listener_ != null) {
					listener_.TRACE(lastMessage_);
				}
				return false;
			} catch (IOException e) {
				lastMessage_ = "ERROR: can't stop feeding data to WebSocket server (" + e.Message + ")";
				if (listener_ != null) {
					listener_.TRACE(lastMessage_);
				}
				return false;
			}
			return true;
		}
	}

	public string getLastMessage() {
		return lastMessage_;
	}

	public int getWaitingResults() {
		return waitingResults_;
	}

	public void sleep(int timeout) {
		try {
			Thread.Sleep(timeout);
		} catch (ThreadInterruptedException) {}
	}

	protected abstract void sendRequest_(byte[] data, int dataOffset, int dataBytes, byte command);

	protected void onOpen_() {
	}

	protected void onClose_() {
	}

	protected void onError_(Exception cause) {
		lock (this) {
			if (state_ == 0 || state_ == 5) {
				return;
			}
			lastMessage_ = "ERROR: caught exception (" + cause.Message + ")";
			if (listener_ != null) {
				listener_.TRACE(lastMessage_);
			}
			state_ = 5;
			Monitor.PulseAll(this);
		}
	}

	protected void onMessage_(string message) {
		if (message.Length == 0) {
			return;
		}
		char command = message[0];
		if (command == 's') {
			if (message.Length > 1) {
				lock (this) {
					lastMessage_ = "ERROR: can't start feeding data to WebSocket server (" + message.Substring(2) + ")";
					if (listener_ != null) {
						listener_.TRACE(lastMessage_);
					}
					state_ = 5;
					Monitor.PulseAll(this);
				}
			} else {
				lock (this) {
					lastMessage_ = "INFO: started feeding data to WebSocket server";
					if (listener_ != null) {
						listener_.TRACE(lastMessage_);
					}
					state_ = 2;
					Monitor.PulseAll(this);
				}
			}
		} else
		if (command == 'p') {
			if (message.Length > 1) {
				lock (this) {
					lastMessage_ = "ERROR: can't feed data to WebSocket server (" + message.Substring(2) + ")";
					if (listener_ != null) {
						listener_.TRACE(lastMessage_);
					}
					state_ = 5;
					Monitor.PulseAll(this);
				}
			} else {
				// pass
			}
		} else
		if (command == 'e') {
			if (message.Length > 1) {
				lock (this) {
					lastMessage_ = "ERROR: can't stop feeding data to WebSocket server (" + message.Substring(2) + ")";
					if (listener_ != null) {
						listener_.TRACE(lastMessage_);
					}
					state_ = 5;
					Monitor.PulseAll(this);
				}
			} else {
				lock (this) {
					lastMessage_ = "INFO: stopped feeding data to WebSocket server";
					if (listener_ != null) {
						listener_.TRACE(lastMessage_);
					}
					state_ = 0;
					Monitor.PulseAll(this);
				}
			}
		} else
		if (command == 'S') {
			waitingResults_++;
			if (listener_ != null) {
				listener_.utteranceStarted(int.Parse(message.Substring(2)));
			}
		} else
		if (command == 'E') {
			if (listener_ != null) {
				listener_.utteranceEnded(int.Parse(message.Substring(2)));
			}
		} else
		if (command == 'C') {
			if (listener_ != null) {
				listener_.resultCreated();
			}
		} else
		if (command == 'U') {
			if (listener_ != null) {
				listener_.resultUpdated(message.Substring(2));
			}
		} else
		if (command == 'A') {
			if (listener_ != null) {
				listener_.resultFinalized(message.Substring(2));
			}
			waitingResults_--;
		} else
		if (command == 'R') {
			if (listener_ != null) {
				listener_.resultFinalized("\x01\x01\x01\x01\x01" + message.Substring(2));
			}
			waitingResults_--;
		} else
		if (command == 'Q') {
			if (listener_ != null) {
				listener_.eventNotified((int)command, message.Substring(2));
			}
		} else
		if (command == 'G') {
			if (listener_ != null) {
				listener_.eventNotified((int)command, message.Substring(2));
			}
		}
	}
}

} // namespace
